### DISCUSSION TITLE
Consecutive 1x1 conv layers

### DISCUSSION THREAD

**Discussion by takekbys** [ultralytics/yolov5#6956] [General]
There are a few consecutive 1x1 conv pairs, I think.
Does it contribute to performance improvement?
Isn't it a wasteful calculation?

 For example, yolov5l has consecutive 1x1 conv pairs.
- end of 8(C3) and beginning of 9(SPPF)
- end of 9(SPPF) and 10(1x1 Conv)
- End of 13(C3) and 14(1x1 Conv)

I'm sorry if I have a misunderstanding.

**Comment by glenn-jocher** [comment]
@takekbys this is normal. Sometimes 1d convolutions are used to reduce channel counts before going into expensive operations, or to match channel counts from other sources, etc.

There's nothing 'wrong' in general with sequential 1D convolutions. This is the basic structure of most smaller fully connected neural networks. Larger kernels like 3x3 are useful for drawing in additional spatial information from the surroundings, which is why they are so common in visual CNNs.

**Reply by takekbys** [reply]
@glenn-jocher Thank you for your comment.

> Sometimes 1d convolutions are used to reduce channel counts before going into expensive operations, or to match channel counts from other sources, etc.

I know the method, and there is no wrong point in the current implementation.
In my understanding, 1D convolutions (beginning of 9 and 10, 14) are inserted in order to reduce channel.
Taking "End of 13(C3) and 14(1x1 Conv)" as an example, 
the same effect can be achieved by reducing only the output channels in C3.cv3.
It costs less, however the difference is very small.

Is there any reason not to take this approach?

**Reply by glenn-jocher** [reply]
@takekbys oh I see what you mean. Yes we could customize the C3 module to separate out `C3.cv3(c1,c2)` `c2` channel count, as right now `c2` affects `c_` which is used throughout the module:
https://github.com/ultralytics/yolov5/blob/2d45de617e0e80fb96424425587b6ce123aa0012/models/common.py#L127-L140

If we don't modify C3, we could also just pass updated c2 values to it, but then the entire C3 module will be doing smaller channel count ops. This may or may not be a good tradeoff (would certainly be faster but mAP may drop), it would need some before and after trainings to quantify the results.

**Reply by glenn-jocher** [reply]
@takekbys I looked a little closer. SPPF() seems like a good candidate to start since c2 there only affects the very last convolution. I've put a TODO here to investigate a few alternatives.
